sources/functions/utils

```diff
@@ -108,3 +108,125 @@ function tmp_swap_off() {
+                        whiptail --title "Error" --msgbox "Passwords did not match." --ok-button="Try again" 10 50
+                    else
+                        echo_warn "Passwords did not match. Press enter to try again"
+                        read -n 1 -r
+                    fi
+                else
+                    #continue if they're good
+                    confirmedPass=true
+                fi
+            fi
+        done
+        unset confirmedPass
+    fi
+    # If password was empty then it needs to be generated
+    if [[ -z "${pass}" ]]; then
+        echo_progress_start "Generating password"
+        pass="$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c16)"
+        whiptail --title "Generated password" --msgbox "The generated password for $user is:\n$pass \n\nNote this down as this will not be shown again after it is hidden" --ok-button="Hide" 10 70
+        echo_progress_done "Password generated and hidden"
+    fi
+
+    #Perform tests. If failure is detected, recurse by unsetting $pass, and running `password_check $1`
+    ##Cracklib test
+    if [[ -n $(which cracklib-check) ]]; then
+        echo_progress_start "Checking password strength"
+        result=$(cracklib-check <<< "$pass")
+        if ! grep -q OK <<< "$result"; then
+            result=""
+            unset result
+            if [[ $1 = "whiptail" ]]; then
+                whiptail --title "Error" --msgbox "Password did not pass cracklib check." --ok-button="Try again" 10 50
+            else
+                echo_warn "Password did not pass cracklib check. Press enter to try again"
+                read -n 1 -r
+            fi
+            pass=
+            unset pass
+            password_check "$1"
+        else
+            echo_progress_done "Password passed"
+        fi
+
+    fi
+}
+
+# Will recurse while $user isn't set or isn't valid
+function username_check() {
+    if [[ -z $user ]]; then
+        if [[ $1 = "whiptail" ]]; then
+            user=$(whiptail --inputbox "Enter username for Swizzin \"master\"" 8 40 3>&1 1>&2 2>&3)
+            exitstatus=$?
+            if [ "$exitstatus" = 1 ]; then exit 1; fi
+        else
+            echo_query "Enter a name for the new user"
+            read -r user
+        fi
+    fi
+
+    #Perform tests. If failure is set message and set flag
+    ## Quick caps check
+    error_message=""
+    if [[ $user =~ [A-Z] ]]; then
+        error_message+="Usernames must not contain capital letters.\n"
+        badUsername=true
+    fi
+    ## Check reserved usernames
+    if [[ $user =~ ("swizzin"|"admin"|"www-data"|"root") ]]; then
+        error_message+="$user is a reserved username -- please use something else.\n"
+        badUsername=true
+    fi
+    ## Do the complicated regex
+    if ! [[ $user =~ ^[a-z_]([a-z0-9_-]{0,31}|[a-z0-9_-]{0,30}\$)$ ]]; then
+        error_message+="Usernames must conform to this regex ^[a-z_]([a-z0-9_-]{0,31}|[a-z0-9_-]{0,30}\\$)$.\n"
```

Comment:
This error message maybe a bit confusing

Reply:
Hahahahaha yeah for sure, just could not figure out how else to word it at the time. Got some suggestions?

Reply:
I'm not entirely sure of the whole string, especially the bit at the end:

Usernames must not contain capitals or start with a number.

Not sure if starting with an underscore will mess anything up further down the line. The main issue is that the sudoers parsing chokes with a number first and a lot of other things break when caps are present.

Reply:
I think this is essentially it?
```suggestion
        error_message+="Usernames must only start with small letters and be up to 32 characters long ^[a-z_]([a-z0-9_-]{0,31}|[a-z0-9_-]{0,30}\\$)$.\n"
```

Reply:
How about:

`Usernames must contain no capitals or special characters, start with a letter and be no longer than 32 characters`

Reply:
Yeah let's roll with that
```suggestion
        error_message+="Usernames must contain no capitals or special characters, start with a letter and be no longer than 32 characters"
```